### PR TITLE
Feature/fsdp lora

### PR DESCRIPTION
**Pull request by danbider**
add a fix to wrap trainable lora modules with fsdp. 
verified successful training on 8 gpus.

related, but not affecting this PR: @bcui19 and I discussed a small enhancement of composer to accompany this PR which will spare big untrained modules from being fetched by fsdp

### CODE REVIEW

**Review by codestar12 — APPROVED** [review]
this looks good I approve to merge if tests pass

**Comment by palash04** [comment]
Hey @danbider , With this PR can we do fsdp lora on llama models? Basically any other model but MPT?

**Comment by danbider** [comment]
> 

should support any model including MPT.

**Comment by alextrott16** [comment]
I'm a fan of the re-design -- in particular moving things into the `hf_causal_lm` builder and out of the `train` script. It makes everything more readable and more general. Also, I really appreciate seeing the addition to the TUTORIAL faqs.

One area that gives me pause is integration with HF for models that have been LoRA-fied. Are there any gotchas when it comes to converting to HF (and uploading to HF) from a Composer checkpoint? Similarly, are there any gotchas when it comes to working with a HF model that is already LoRA-fied?

My sense is that, with the latter, everything should be OK as long as the right things are installed, but I'd like to get a sanity check on that.

The former seems like it will still be missing support. @dakinggg can probably add some insight here, because I'm worried that the code that actually modifies the model (in `hf_causal_lm`) won't be reflected in the model construction code that gets uploaded to the HF repo along with the model weights. Has anyone tested that workflow?

File: llmfoundry/models/hf/hf_causal_lm.py [hunk 1]
```diff
@@ -76,108 +87,106 @@ def __init__(self, om_model_config: _om_model_config_type,
...
+            if _peft_installed == True:
+                print('Building Lora config...')
+                lora_cfg = LoraConfig(**lora_cfg.args)
+                print('Lora config built.')
+                print('Adding Lora modules...')
+                model = get_peft_model(model, lora_cfg)
+                print('Lora modules added.')
+                print_trainable_parameters(model)
+            else:
+                warn(
```
Comment (alextrott16) [inline llmfoundry/models/hf/hf_causal_lm.py:185]: I'd vote for this to be an error rather than a warning.

Comment (samhavens) [reply llmfoundry/models/hf/hf_causal_lm.py:185]: me too

File: TUTORIAL.md [hunk 2]
```diff
@@ -330,16 +330,41 @@ The majority of our training setups use `triton`. -->
...
+  lora:
+    args:
+      r: 16
+      lora_alpha: 32
+      target_modules: ["Wqkv", "out_proj", "up_proj", "down_proj"]
+      lora_dropout: 0.05
+      bias: none
+      task_type: "CAUSAL_LM"
+```
+You can train LoRA models either using FSDP for further memory savings. in your `.yaml`, specify:
```
Comment (dakinggg) [inline TUTORIAL.md:349]:
```suggestion
You can train LoRA models using FSDP for further memory savings. in your `.yaml`, specify:
```

File: TUTORIAL.md [hunk 3]
```diff
@@ -330,16 +330,41 @@ The majority of our training setups use `triton`. -->
...
+<!--pytest.mark.skip-->
+```yaml
+fsdp_config:
+  use_orig_params: true
```
Comment (dakinggg) [inline TUTORIAL.md:353]: Can we confirm if this is necessary?

File: TUTORIAL.md [hunk 4]
```diff
@@ -330,16 +330,41 @@ The majority of our training setups use `triton`. -->
...
+  sharding_strategy: FULL_SHARD
+  mixed_precision: PURE
+  activation_checkpointing: true
+  activation_checkpointing_reentrant: false
+  activation_cpu_offload: false
+  limit_all_gathers: true
 ```
+or default to DDP, as follows:
```
Comment (dakinggg) [inline TUTORIAL.md:361]: I think to default DDP just leaving out the FSDP section entirely is a bit cleaner?

File: llmfoundry/models/hf/hf_fsdp.py [hunk 5]
```diff
@@ -190,6 +198,17 @@ def prepare_hf_causal_lm_model_for_fsdp(model: PreTrainedModel,
             tied_embeddings._fsdp_wrap = False  # type: ignore
             lm_head._fsdp_wrap = False  # type: ignore
 
+    # applying ._fsdp_wrap = True for the LoRA modules
+    # this is needed because added LoRA modules have requires_grad=True,
+    # while the rest of the modules have requires_grad=False
+    if lora_model_type is not None:  # peft is installed
+        if isinstance(model.base_model,
+                      lora_model_type):  # we have builR a LoraModel
```
Comment (dakinggg) [inline llmfoundry/models/hf/hf_fsdp.py:206]:
```suggestion
                      lora_model_type):  # we have built a LoraModel
```

File: scripts/train/train.py [hunk 6]
```diff
@@ -226,10 +183,14 @@ def main(cfg: DictConfig):
     fsdp_config: Optional[Dict] = om.to_container(
         fsdp_dict_config
     ) if fsdp_dict_config is not None else None  # type: ignore
-    lora_config: Optional[DictConfig] = pop_config(cfg,
+    lora_config: Optional[DictConfig] = pop_config(model_config,
                                                    'lora',
                                                    must_exist=False,
                                                    default_value=None)
+    if lora_config is not None:
+        if lora_config.get('rank', None) is not None:
```
Comment (dakinggg) [inline scripts/train/train.py:191]: What is supposed to happen if lora config is provided but rank is none? Should that be an error?

**Review by dakinggg — COMMENTED** [review]
Could you please also add some basic tests for the lora addition?

File: TUTORIAL.md (hunk 3 above)
Comment (danbider) [reply TUTORIAL.md:353]: will verify this tomorrow AM, good point

**Comment by danbider** [comment]
> Could you please also add some basic tests for the lora addition?

will add those now.

**Comment by samhavens** [comment]
Any chance we could include `scripts/train/yamls/finetune/lora-mpt-7b.yaml`

```yaml
max_seq_len: 2048
global_seed: 17
dist_timeout: 5400

# Run Name
run_name: # If left blank, will be read from env var $RUN_NAME

model:
  name: hf_causal_lm
  pretrained: true
  pretrained_model_name_or_path: mosaicml/mpt-7b
  config_overrides:
    attn_config:
      attn_impl: triton
      attn_uses_sequence_id: false
  lora:
    # UPDATE these as needed
    args:
      r: 1
      lora_alpha: 32
      # target_modules: ["Wqkv", "out_proj", "up_proj", "down_proj"]
      target_modules: ["up_proj", "down_proj"]
      lora_dropout: 0.05
      bias: none
      task_type: "CAUSAL_LM"
# Tokenizer
tokenizer:
  name: mosaicml/mpt-7b
  kwargs:
    model_max_length: ${max_seq_len}

# Dataloaders
train_loader:
  name: finetuning
  dataset:
    hf_name: danbider/codegen
    split: train
    max_seq_len: ${max_seq_len}
    allow_pad_trimming: false
    decoder_only_format: true
    packing_ratio: 19.6 # concat examples
    shuffle: true
  drop_last: true
  num_workers: 8
  pin_memory: false
  prefetch_factor: 2
  persistent_workers: true
  timeout: 0

eval_loader:
  name: finetuning
  dataset:
    hf_name: danbider/codegen
    split: test
    max_seq_len: ${max_seq_len}
    allow_pad_trimming: false
    decoder_only_format: true
    packing_ratio: 19.6
    shuffle: true
  drop_last: true
  num_workers: 8
  pin_memory: false
  prefetch_factor: 2
  persistent_workers: true
  timeout: 0

# Optimization
# Based on MPT pretraining
scheduler:
  name: cosine_with_warmup
  t_warmup: 50ba
  alpha_f: 0.0

optimizer:
  name: decoupled_lionw
  lr: 1.0e-4  # lora needs higher LR
  betas:
  - 0.9
  - 0.95
  eps: 1.0e-8
  weight_decay: 0

algorithms:
  gradient_clipping:
    clipping_type: norm
    clipping_threshold: 1.0

max_duration: 2ep
eval_interval: 1ep
eval_first: true
global_train_batch_size: 48

# System
seed: ${global_seed}
device_eval_batch_size: 4
device_train_microbatch_size: 1
precision: amp_bf16

# FSDP
# fsdp_config:
#   sharding_strategy: FULL_SHARD
#   mixed_precision: PURE
#   activation_checkpointing: true
#   activation_checkpointing_reentrant: false
#   activation_cpu_offload: false
#   limit_all_gathers: true
#   verbose: false

# leave out fsdp_config for DDP or single GPU

# Logging
progress_bar: false
log_to_console: true
console_log_interval: 1ba

callbacks:
  speed_monitor:
    window_size: 10
  lr_monitor: {}
  memory_monitor: {}
  runtime_estimator: {}

# loggers:
#   wandb: {}
# Checkpoint to local filesystem or remote object store
save_interval: 5000ba
save_num_checkpoints_to_keep: 1  # Important, this cleans up checkpoints saved to DISK
save_folder: ./{run_name}/checkpoints
# save_folder: s3://my-bucket/my-folder/{run_name}/checkpoint
```

Any example we could point users to would be great, even if we plan on refining it later

**Comment by germanjke** [comment]
Hi, i tested your branch and got some bugs:

1. Bug about build Lora (we can easy fix this)
2. Bug about wrap FSDP Lora (need to fix)

I'm using your suggestion from `TUTORIAL.md`:

I'm using LLaMA-2 LorA

```
model:
  name: hf_causal_lm
  pretrained: true
  ...
  lora:
    args:
      r: 16
      lora_alpha: 32
      target_modules: ["Wqkv", "out_proj", "up_proj", "down_proj"] # or any subset of these for MPT-7B
      lora_dropout: 0.05
      bias: none
      task_type: "CAUSAL_LM"
```

```
fsdp_config:
  use_orig_params: true
  sharding_strategy: FULL_SHARD
  mixed_precision: PURE
  activation_checkpointing: true
  activation_checkpointing_reentrant: false
  activation_cpu_offload: false
  limit_all_gathers: true
```

So, let's start with bug №1:

You define `model_config` [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/scripts/train/train.py#L170), everything is ok, we have lora section there. 

[Here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/scripts/train/train.py#L352) you build the model, but lora section alraedy not here. So, `lora_cfg` is `None` [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_causal_lm.py#L178).

It happens, because you poping this lora section here and not using anywhere, only using for `print` [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/scripts/train/train.py#L192)

`main` branch have [this](https://github.com/danbider/llm-foundry/blob/main/scripts/train/train.py#L73) function, and they building model with 2 different configs: `model_config` and `lora_cfg`, you building this with single `model_config`, but `LoRA` popped from there, so we dont build `LoRA` here.

You need not to pop, or use function from `main` branch. 
After this `LoRA` builds. 

About bug №2:

We build `LoRA` [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_causal_lm.py#L185)

Later, you making reinit of your LoRA model, so we have error [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_causal_lm.py#L208) cause of FSDP [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/model_wrapper.py#L64)

For `LLaMA 2` we can get [this](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_fsdp.py#L102), but we can't get this for `LLaMA 2 LoRA`, cause of this we have `None` [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_fsdp.py#L147) and we raising [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_fsdp.py#L161)

I think we need to rename [this](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_fsdp.py#L102C11-L102C33) for Lora Llama maybe or refactor this in other way

Thanks! I hope soon we will can train `LoRA` models with FSDP 👍

**Comment by danbider** [comment]
> Hi, i tested your branch and got some bugs:
> 
> 1. Bug about build Lora (we can easy fix this)
> 2. Bug about wrap FSDP Lora (need to fix)
> 
> I'm using your suggestion from `TUTORIAL.md`:
> 
> I'm using LLaMA-2 LorA
> 
> ```
> model:
>   name: hf_causal_lm
>   pretrained: true
>   ...
>   lora:
>     args:
>       r: 16
>       lora_alpha: 32
>       target_modules: ["Wqkv", "out_proj", "up_proj", "down_proj"] # or any subset of these for MPT-7B
>       lora_dropout: 0.05
>       bias: none
>       task_type: "CAUSAL_LM"
> ```
> 
> ```
> fsdp_config:
>   use_orig_params: true
>   sharding_strategy: FULL_SHARD
>   mixed_precision: PURE
>   activation_checkpointing: true
>   activation_checkpointing_reentrant: false
>   activation_cpu_offload: false
>   limit_all_gathers: true
> ```
> 
> So, let's start with bug №1:
> 
> You define `model_config` [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/scripts/train/train.py#L170), everything is ok, we have lora section there.
> 
> [Here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/scripts/train/train.py#L352) you build the model, but lora section alraedy not here. So, `lora_cfg` is `None` [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_causal_lm.py#L178).
> 
> It happens, because you poping this lora section here and not using anywhere, only using for `print` [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/scripts/train/train.py#L192)
> 
> `main` branch have [this](https://github.com/danbider/llm-foundry/blob/main/scripts/train/train.py#L73) function, and they building model with 2 different configs: `model_config` and `lora_cfg`, you building this with single `model_config`, but `LoRA` popped from there, so we dont build `LoRA` here.
> 
> You need not to pop, or use function from `main` branch. After this `LoRA` builds.
> 
> About bug №2:
> 
> We build `LoRA` [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_causal_lm.py#L185)
> 
> Later, you making reinit of your LoRA model, so we have error [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_causal_lm.py#L208) cause of FSDP [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/model_wrapper.py#L64)
> 
> For `LLaMA 2` we can get [this](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_fsdp.py#L102), but we can't get this for `LLaMA 2 LoRA`, cause of this we have `None` [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_fsdp.py#L147) and we raising [here](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_fsdp.py#L161)
> 
> I think we need to rename [this](https://github.com/danbider/llm-foundry/blob/feature/fsdp-lora/llmfoundry/models/hf/hf_fsdp.py#L102C11-L102C33) for Lora Llama maybe or refactor this in other way
> 
> Thanks! I hope soon we will can train `LoRA` models with FSDP 👍

thanks for this. fixed the first one, we think. will take care of the second as well.

**Comment by dakinggg** [comment]
Closing in favor of https://github.com/mosaicml/llm-foundry/pull/886
